src/main/java/graphql/schema/idl/SchemaPrinter.java

```diff
@@ -1041,25 +1041,50 @@ private boolean hasDeprecatedDirective(List<GraphQLAppliedDirective> directives)
                 .count() == 1;
     }
 
-    private List<GraphQLAppliedDirective> addDeprecatedDirectiveIfNeeded(GraphQLDirectiveContainer directiveContainer) {
+    private List<GraphQLAppliedDirective> addOrUpdateDeprecatedDirectiveIfNeeded(GraphQLDirectiveContainer directiveContainer) {
         List<GraphQLAppliedDirective> directives = DirectivesUtil.toAppliedDirectives(directiveContainer);
+        String reason = getDeprecationReason(directiveContainer);
+
         if (!hasDeprecatedDirective(directives) && isDeprecatedDirectiveAllowed()) {
             directives = new ArrayList<>(directives);
-            String reason = getDeprecationReason(directiveContainer);
-            GraphQLAppliedDirectiveArgument arg = GraphQLAppliedDirectiveArgument.newArgument()
-                    .name("reason")
-                    .valueProgrammatic(reason)
-                    .type(GraphQLString)
-                    .build();
-            GraphQLAppliedDirective directive = GraphQLAppliedDirective.newDirective()
-                    .name("deprecated")
-                    .argument(arg)
-                    .build();
-            directives.add(directive);
+            directives.add(createDeprecatedDirective(reason));
+        } else if (hasDeprecatedDirective(directives) && isDeprecatedDirectiveAllowed()) {
+            // Update deprecated reason in case modified by schema transform
+            directives = updateDeprecatedDirective(directives, reason);
         }
         return directives;
     }
 
+    private GraphQLAppliedDirective createDeprecatedDirective(String reason) {
+        GraphQLAppliedDirectiveArgument arg = GraphQLAppliedDirectiveArgument.newArgument()
+                .name("reason")
+                .valueProgrammatic(reason)
+                .type(GraphQLString)
+                .build();
+        return GraphQLAppliedDirective.newDirective()
+                .name("deprecated")
+                .argument(arg)
+                .build();
+    }
+
+    private List<GraphQLAppliedDirective> updateDeprecatedDirective(List<GraphQLAppliedDirective> directives, String reason) {
+        GraphQLAppliedDirectiveArgument newArg = GraphQLAppliedDirectiveArgument.newArgument()
+                .name("reason")
+                .valueProgrammatic(reason)
+                .type(GraphQLString)
+                .build();
+
+        return directives.stream().map(d -> {
+            if (isDeprecatedDirective(d)) {
+                // Don't include reason is deliberately replaced with NOT_SET, for example in Anonymizer
+                if (d.getArgument("reason").getArgumentValue() != InputValueWithState.NOT_SET) {
```

Comment:
This is a very special case. The Anonymizer wipes out deprecated reasons with a special `NOT_SET` value, which is distinct from `null`. If this is spotted, we won't add back the deprecated reason.